---
title: "TCGA-BLCA"
author: "Himanshu Chintalapudi"
date: "5/11/2022"
output: html_document
editor_options: 
  chunk_output_type: console
---

```{r}
source("libs.R")
```


```{r}
blca_metadata<- read.delim("/Users/hchintalapudi/Desktop/work/TCGA analysis/PanCancer/TCGA-BLCA/BLCA_gdc_sample_sheet.2021-11-11.tsv")

blca_metadata<- blca_metadata[!blca_metadata$File.Name %in% c("5b802329-910c-4716-82ad-72362793bcd7_gdc_realn_rehead.bam", "62c259cc-7b78-41df-bec3-c071ac8d3017_gdc_realn_rehead.bam",
"63251856-dd47-4dd8-83c8-b32341f38a09_gdc_realn_rehead.bam","da2df068-9ab9-48d5-8bb1-9e48d4b77e50_gdc_realn_rehead.bam","e087ebc1-7d99-4228-a447-b0c62e64003b_gdc_realn_rehead.bam"),]

blca_metadata$File.Name<- gsub(x = blca_metadata$File.Name, pattern = ".bam", replacement = "")
samples<- as.character(blca_metadata$File.Name)


#--- Final Counts from Telescope:
blca_counts.telescope.tmp <- lapply(samples,
                           function(s){
                               tmp <- read.table(
                                   file.path('/Users/hchintalapudi/Desktop/work/TCGA analysis/PanCancer/TCGA-BLCA/Telescope/',paste0(s, '_sorted-telescope_report.tsv')),
                                   sep='\t', header=T, stringsAsFactors=F)
                               ret <- data.frame(transcript=genes$transcript_id, stringsAsFactors=F) %>%
                                   left_join(tmp, by='transcript') %>%
                                   dplyr::mutate(
                                       gene_id = transcript,
                                       count = final_count
                                   ) %>%
                                   dplyr::select(gene_id, count)
                               ret[is.na(ret)] <- 0
                               stopifnot(all(ret$gene_id == genes$transcript_id))
                               #ret$gene_id <- NULL
                               names(ret) <- c(s)
                               ret
                           }) %>%
    bind_cols
rownames(blca_counts.telescope)<- blca_counts.telescope.tmp$`c54f46b6-eebf-4433-8e61-699fd37ed65f_gdc_realn_rehead`
table(blca_metadata$File.Name == colnames(blca_counts.telescope))

colnames(blca_counts.telescope)<- blca_metadata$Sample.ID
write.csv(blca_counts.telescope, file = "/Users/hchintalapudi/Desktop/work/TCGA analysis/PanCancer/TCGA-BLCA_Telescope_finalcounts.csv")
blca_counts.telescope<- read.csv("/Users/hchintalapudi/Desktop/work/TCGA analysis/PanCancer/TCGA-BLCA_Telescope_finalcounts.csv", check.names = F, stringsAsFactors = F, row.names = 1)
## subset to:
blca_counts.telescope.2<-blca_counts.telescope
blca_counts.telescope.2<- as.data.frame(t(blca_counts.telescope.2))
rownames(blca_counts.telescope.2)<- gsub("\\.","-", rownames(blca_counts.telescope.2))
blca_counts.telescope.2<-blca_counts.telescope.2%>%
  mutate(tmp1 = sapply(strsplit(rownames(blca_counts.telescope.2), "-"), function(x) x[[4]][[1]]))%>%
  mutate(tmp2 = gsub("-01A|-01B","",rownames(blca_counts.telescope.2)))
newcolumn_blca_counts = blca_counts.telescope.2 %>%
  dplyr::select(tmp1,tmp2)%>%
  group_by(tmp2)%>%
  mutate(tmp3 = paste0(tmp1,collapse = "-"))%>%
  dplyr::slice(1)%>%
  dplyr::select(-tmp1)%>%
  mutate(new_name = paste(tmp2,tmp3,sep = "-",collapse = "-"))

blca_counts.telescope.2=blca_counts.telescope.2%>%
  dplyr::select(-tmp1)%>%
  group_by(tmp2)%>%summarise_all(funs(mean))

blca_counts.telescope.2<- left_join(blca_counts.telescope.2, newcolumn_blca_counts[,c(1,3)], by ="tmp2")
blca_counts.telescope.2<- blca_counts.telescope.2%>%
  dplyr::select(-tmp2)%>%
  column_to_rownames("new_name")

blca_counts.telescope.2<- as.data.frame(t(blca_counts.telescope.2))
blca_counts.telescope.2<- blca_counts.telescope.2[colnames(blca_counts.telescope.2) %in% colnames(blca_genecounts_am_tpm_filt)]

write.csv(blca_counts.telescope.2, file = "/Users/hchintalapudi/Desktop/work/TCGA analysis/PanCancer/TCGA-BLCA_Telescope_finalcounts_filt.csv")


## SIZE Factors for DESeq2:
metrics.list <- lapply(samples, 
                       function(s){
                           f <- file.path('/Users/hchintalapudi/Desktop/work/TCGA analysis/PanCancer/TCGA-BLCA/Telescope/',paste0(s, '_sorted-telescope_report.tsv'))
                           if(file.exists(f)) {
                               print(f)
                               h <- readLines(f, 1) %>% strsplit(., '\t') %>% unlist
                               rstr <- sapply(strsplit(h[-c(1,2)], ':'), function(t) as.numeric(unlist(t[2][1])))
                               names(rstr) <- sapply(strsplit(h[-c(1,2)], ':'), function(t) t[1])
                           } else{
                               rstr <- c(NA)
                           }
                           rstr
                       }
)
mn <- unique(do.call(c, lapply(metrics.list, names)))
met.ts <- lapply(metrics.list, 
                 function(m) {
                     ret <- sapply(mn, function(x) m[x])
                     names(ret) <- gsub('\\..*', '', names(ret))
                     ret
                 }) %>% do.call(rbind, .) %>% data.frame
names(met.ts) <- paste(names(met.ts), 'ts', sep='.')
met.ts$sample<- samples
blca_metrics<- met.ts
blca_metrics$sample
blca_metrics<- left_join(blca_metrics, blca_metadata[,c(1, 2,6,7)], by = c("sample"="File.Name"))

#--- Add library size (total frags mapped) columns
blca_metrics$mapped_frags.bt2 <- blca_metrics$pair_mapped.ts + blca_metrics$pair_mixed.ts + blca_metrics$single_mapped.ts
write.csv(blca_metrics, file = "/Users/hchintalapudi/Desktop/work/TCGA analysis/PanCancer/TCGA-BLCA/blca_telescope_metrics.csv")

# Function for calculating size factors
calculateSizeFactors <- function(mapped_frags) {
    geomean <- expm1(mean(log1p(mapped_frags)))
    mapped_frags / geomean
}
blca_sizefactors<- as.data.frame(calculateSizeFactors(blca_metrics$mapped_frags.bt2))
blca_sizefactors$Sample<- blca_metadata$Sample.ID
blca_sizefactors<- blca_sizefactors[,c(2,1)]
colnames(blca_sizefactors)<- c("Sample","sizefactor")
write.csv(blca_sizefactors, file = "/Users/hchintalapudi/Desktop/work/TCGA analysis/PanCancer/TCGA-BLCA/Telescope_skcm_sizefactors.csv", row.names = F)

blca_metadata.2<- read.delim("/Users/hchintalapudi/Desktop/work/TCGA analysis/PanCancer/TCGA-BLCA/gdc_download_20210913_160149.716118/eaa71705-960a-4abd-b5d7-f5fdc0d0c5af/nationwidechildrens.org_clinical_patient_blca.txt")
blca_metadata.2<- blca_metadata.2[-c(1,2),]
blca_metadata.2<- blca_metadata.2[,c("bcr_patient_barcode","vital_status","age_at_diagnosis", "ajcc_pathologic_tumor_stage", "last_contact_days_to","death_days_to")]
blca_metadata.2<- left_join(blca_metadata.2, blca_metadata[,c(6,7,8)], by = c("bcr_patient_barcode" = "Case.ID"))
blca_metadata.2<- blca_metadata.2[blca_metadata.2$Sample.Type %in% "Primary Tumor",]

set.seed(12345)
blca_metadata.2$survival.time <- as.integer(ifelse(blca_metadata.2$last_contact_days_to == "[Not Available]",
                                   blca_metadata.2$death_days_to, blca_metadata.2$last_contact_days_to))
blca_metadata.2$survival.time[blca_metadata.2$survival.time<0]<-0
blca_metadata.2$ajcc_pathologic_tumor_stage[blca_metadata.2$ajcc_pathologic_tumor_stage == "[Not Available]"]<- NA
blca_metadata.2<- blca_metadata.2%>%
  mutate(censor = ifelse(vital_status == "Alive", 0, ifelse(vital_status == "Dead", 1, NA)))
blca_metadata.2<- blca_metadata.2[,c(7,9,10,3,4)]
blca_metadata.2[duplicated(blca_metadata.2$Sample.ID),]
blca_metadata.2<- unique(blca_metadata.2)

## get vst counts:
blca_metadata_deseq<- blca_metadata.2[,c(1,2)]
blca_metadata_deseq<- blca_metadata_deseq[complete.cases(blca_metadata_deseq),]
rownames(blca_metadata_deseq)<- NULL
median(blca_metadata_deseq$survival.time)
blca_metadata_deseq<- blca_metadata_deseq%>%
  mutate(Survival = ifelse(survival.time>=259, "LTS", ifelse(survival.time<=259, "STS",NA)))%>%
  column_to_rownames("Sample.ID")%>%
  dplyr::select(-1)

blca_counts.telescope_filt<- blca_counts.telescope[rownames(blca_metadata_deseq)]
table(colnames(blca_counts.telescope_filt) == rownames(blca_metadata_deseq))

dds_blca<- DESeqDataSetFromMatrix(countData = blca_counts.telescope_filt, colData =blca_metadata_deseq, design = ~Survival)
dds_blca$Survival<- factor(dds_blca$Survival, levels = c("STS", "LTS"))
blca_sizefactors<- blca_sizefactors[match(rownames(blca_metadata_deseq), blca_sizefactors$Sample),]
table(rownames(blca_metadata_deseq) == blca_sizefactors$Sample)

sizeFactors(dds_blca)<- blca_sizefactors$sizefactor
dds_blca.2<- dds_blca[rowSums(counts(dds_blca)>=10)>=2,]
table(rowSums(counts(dds_blca)>=10)>=2)

blca_vst<- as.data.frame(assay(vst(dds_blca.2, blind = F)))
write.csv(blca_vst, file = "/Users/hchintalapudi/Desktop/work/TCGA analysis/PanCancer/TCGA-BLCA/blca_vst_9933.csv")
blca_vst<- read.csv("/Users/hchintalapudi/Desktop/work/TCGA analysis/PanCancer/TCGA-BLCA/blca_vst_9933.csv", row.names = 1, check.names = F)

blca_metadata.2.1<- blca_metadata.2[blca_metadata.2$Sample.ID %in% colnames(blca_vst),]
table(colnames(blca_vst) == blca_metadata.2.1$Sample.ID)

blca_cox_input<- data.frame(blca_metadata.2.1[,c("Sample.ID","survival.time","censor","age_at_diagnosis", "ajcc_pathologic_tumor_stage")], t(blca_vst))
rownames(blca_cox_input)<- NULL
blca_cox_input<- blca_cox_input%>%
  column_to_rownames("Sample.ID")

blca_cox_input<- blca_cox_input[complete.cases(blca_cox_input),]

blca_cox_input[,c(1,2,3)]<- lapply(blca_cox_input[,c(1,2,3)], as.numeric)
blca_cox_input$ajcc_pathologic_tumor_stage<- as.factor(blca_cox_input$ajcc_pathologic_tumor_stage)


blca_samples_filt_Ali<- read.table("/Users/hchintalapudi/Desktop/work/TCGA analysis/PanCancer/TCGA-BLCA/case_set_BLCA_ERV__no_prior_tx.2021-11-18.tsv", header = T)
blca_metadata.2<- blca_metadata.2%>%
  mutate(bcr_patient_uuid = tolower(bcr_patient_uuid))
blca_samples_filt_Ali<- left_join(blca_samples_filt_Ali, blca_metadata.2[,c(1,2)], by = c("id" = "bcr_patient_uuid"))

blca_metadata.trim<- blca_metadata
blca_metadata.trim<- blca_metadata.trim[!grepl("-11A|-02", blca_metadata.trim$Sample.ID),]
blca_metadata.trim<- blca_metadata.trim%>%
  mutate(tmp1 = sapply(strsplit(Sample.ID, "-"), function(x) x[[4]][[1]]))%>%
  mutate(tmp2 = gsub("-01A|-01B","",Sample.ID))%>%
  group_by(tmp2)%>%filter(!duplicated(tmp1))



blca_samples_filt_Ali<- left_join(blca_samples_filt_Ali, blca_metadata.trim[,c(6,7)], by = c("bcr_patient_barcode" = "Case.ID"))

tmp<-blca_samples_filt_Ali[!duplicated(blca_samples_filt_Ali$id),]

blca_samples_filt_Ali<- read.csv("/Users/hchintalapudi/Desktop/work/TCGA analysis/PanCancer/TCGA-BLCA/case_set_BLCA_ERV__no_prior_tx.2021-11-18_updated.tsv", header = T, row.names = 1)
samples<- as.character(blca_samples_filt_Ali$File.Name)

blca_cox_input<- blca_cox_input[rownames(blca_cox_input) %in% blca_samples_filt_Ali$Sample.ID,]

## Cox with RegParalell:
# ** 01/05/22: plugging in the new input file where cases are filtered and the 01As and 01Bs are averaged to be a single sample:
blca_cox_res<-RegParallel(
    data = blca_cox_input.2.1,
    formula = 'Surv(survival.time, censor) ~ [*] + age_at_diagnosis + ajcc_pathologic_tumor_stage',
    FUN = function(formula, data)
      coxph(formula = formula,
        data = data,
        ties = 'breslow',
        singular.ok = TRUE),
    FUNtype = 'coxph',
    variables = colnames(blca_cox_input.2.1)[5:ncol(blca_cox_input.2.1)],
    blocksize = 3000,
    cores = 2,
    nestedParallel = FALSE,
    conflevel = 95,
    excludeTerms = c("age_at_diagnosis", "ajcc_pathologic_tumor_stage"),
    excludeIntercept = TRUE,
    p.adjust = 'fdr')
write.csv(blca_cox_input, file = "/Users/hchintalapudi/Desktop/work/TCGA analysis/PanCancer/TCGA-BLCA/blca_cox_input.csv")
## old unfiltered file, use with caution!! 
blca_cox_input<- read.csv("/Users/hchintalapudi/Desktop/work/TCGA analysis/PanCancer/TCGA-BLCA/blca_cox_input.csv", row.names = 1, check.names = F)
## * 01/05/2022:
write.csv(blca_cox_res, file = "/Users/hchintalapudi/Desktop/work/TCGA analysis/PanCancer/TCGA-BLCA/BLCA-vst_cox_results.csv")
blca_cox_res<- read.csv("/Users/hchintalapudi/Desktop/work/TCGA analysis/PanCancer/TCGA-BLCA/BLCA-vst_cox_results.csv", row.names = 1)

blca_unstr<- read.delim("/Users/hchintalapudi/Desktop/work/TCGA analysis/PanCancer/TCGA-BLCA/TCGA-BLCA_unstr-counts_390.txt", check.names = F, stringsAsFactors = F, row.names = 1)
blca_unstr<- blca_unstr[samples]
table(samples == colnames(blca_unstr))

blca_unstr<- blca_unstr[samples]
table(blca_samples_filt_Ali$File.Name == samples)
colnames(blca_unstr)<- blca_samples_filt_Ali$Sample.ID
write.csv(blca_unstr, file = "/Users/hchintalapudi/Desktop/work/TCGA analysis/PanCancer/TCGA-BLCA/TCGA-BLCA_raw_genecounts_390.txt")
```


```{r}
##TPM:
tpm3 <- function(counts,len) {
  x <- counts/len
  return(t(t(x)*1e6/colSums(x)))
}

blca_genecounts<- read.csv("/Users/hchintalapudi/Desktop/work/TCGA analysis/PanCancer/TCGA-BLCA/TCGA-BLCA_raw_genecounts_390.txt", row.names = 1, check.names = F)

blca_genecounts_am_filt<-blca_genecounts_am[blca_samples_filt_Ali$Sample.ID]
blca_genecounts_am_filt<- as.data.frame(t(blca_genecounts_am_filt))
blca_genecounts_am_filt<-blca_genecounts_am_filt%>%
  mutate(tmp1 = sapply(strsplit(rownames(blca_genecounts_am_filt), "-"), function(x) x[[4]][[1]]))%>%
  mutate(tmp2 = gsub("-01A|-01B","",rownames(blca_genecounts_am_filt)))
newcolumn = blca_genecounts_am_filt %>%
  dplyr::select(tmp1,tmp2)%>%
  group_by(tmp2)%>%
  mutate(tmp3 = paste0(tmp1,collapse = "-"))%>%
  dplyr::slice(1)%>%
  dplyr::select(-tmp1)%>%
  mutate(new_name = paste(tmp2,tmp3,sep = "-",collapse = "-"))


blca_genecounts_am_filt=blca_genecounts_am_filt%>%
  dplyr::select(-tmp1)%>%
  group_by(tmp2)%>%summarise_all(funs(mean))

blca_genecounts_am_filt<- left_join(blca_genecounts_am_filt, newcolumn[,c(1,3)], by ="tmp2")
blca_genecounts_am_filt<- blca_genecounts_am_filt%>%
  dplyr::select(-tmp2)%>%
  column_to_rownames("new_name")

gene_lengths<- exonic.gene.sizes
gene_lengths<- gene_lengths[!duplicated(gene_lengths$gene_name),]
table(gene_lengths$gene_name %in% rownames(blca_genecounts_am))

gene_lengths<- gene_lengths[match(rownames(blca_genecounts_am), gene_lengths$gene_name),]
colnames(gene_lengths)[2]<- "gene_length"
table(gene_lengths$gene_name == rownames(blca_genecounts_am))


blca_genecounts_tpm<- as.data.frame(tpm3((blca_genecounts),gene_lengths$gene_length))
write.csv(blca_genecounts_tpm, file = "/Users/hchintalapudi/Desktop/work/TCGA analysis/PanCancer/TCGA-BLCA/BLCA_tpm_390.csv")

write.csv(gene_lengths, file = "/Users/hchintalapudi/Desktop/work/TCGA analysis/PanCancer/gencode.v31.genelengths.csv")
gene_lengths<- read.csv("/Users/hchintalapudi/Desktop/work/TCGA analysis/PanCancer/gencode.v31.genelengths.csv", row.names = 1)


blca_genecounts_am_tpm<- read.csv("/Users/hchintalapudi/Desktop/work/TCGA analysis/PanCancer/TCGA-BLCA/blca_genecounts_AM_tpm.csv", check.names = F, row.names = 1)
blca_genecounts_am_filt<- as.data.frame(t(blca_genecounts_am_filt))

write.csv(blca_genecounts_am_tpm_filt, file = "/Users/hchintalapudi/Desktop/work/TCGA analysis/PanCancer/TCGA-BLCA/blca_genecounts_AM_tpm.csv")
blca_genecounts_am_tpm_filt<- read.csv("/Users/hchintalapudi/Desktop/work/TCGA analysis/PanCancer/TCGA-BLCA/blca_genecounts_AM_tpm.csv", row.names = 1, check.names = F)


blca_cox_input.2<-blca_cox_input
blca_cox_input.2<-blca_cox_input.2%>%
  mutate(tmp1 = sapply(strsplit(rownames(blca_cox_input.2), "-"), function(x) x[[4]][[1]]))%>%
  mutate(tmp2 = gsub("-01A|-01B","",rownames(blca_cox_input.2)))
newcolumn_blca = blca_cox_input.2 %>%
  dplyr::select(tmp1,tmp2,survival.time, censor, age_at_diagnosis, ajcc_pathologic_tumor_stage)%>%
  group_by(tmp2)%>%
  mutate(tmp3 = paste0(tmp1,collapse = "-"))%>%
  dplyr::slice(1)%>%
  dplyr::select(-tmp1)%>%
  mutate(new_name = paste(tmp2,tmp3,sep = "-",collapse = "-"))


blca_cox_input.2=blca_cox_input.2[,-c(1:4)]%>%
  dplyr::select(-tmp1)%>%
  group_by(tmp2)%>%summarise_all(funs(mean))

blca_cox_input.2<- left_join(blca_cox_input.2, newcolumn_blca, by ="tmp2")
blca_cox_input.2<- blca_cox_input.2%>%
  dplyr::select(-c(tmp2,tmp3))%>%
  column_to_rownames("new_name")
blca_cox_input.2<- blca_cox_input.2[,c(9934:9937,1:9933)]
table(rownames(blca_cox_input.2) %in% colnames(blca_genecounts_am_tpm_filt))
blca_cox_input.2.1<- blca_cox_input.2[rownames(blca_cox_input.2) %in% colnames(blca_genecounts_am_tpm_filt),]
write.csv(blca_cox_input.2.1, file = "/Users/hchintalapudi/Desktop/work/TCGA analysis/PanCancer/TCGA-BLCA/blca_cox_input_filt.csv")
```

