---
title: "TCGA-OV"
author: "Himanshu Chintalapudi"
date: "5/11/2022"
output: html_document
---

```{r}
## TCGA-OV:
OV_metadata<- read.table("/Users/hchintalapudi/Desktop/work/TCGA analysis/PanCancer/TCGA-OV/TCGA-OV_gdc_sample_sheet.2021-09-01.tsv", header = T, sep = "\t")
OV_metadata<- OV_metadata[!grepl("-11A|-02", OV_metadata$Sample.ID),]

OV_metadata$File.Name<- gsub(x = OV_metadata$File.Name, pattern = ".bam", replacement = "")
#******Removing "bcdf0550-9e6f-4500-9473-706df7beeae4_gdc_realn_rehead" as the fastq was truncated
OV_metadata<- OV_metadata[!OV_metadata$File.Name %in% "bcdf0550-9e6f-4500-9473-706df7beeae4_gdc_realn_rehead",]
samples<- as.character(OV_metadata$File.Name)

#--- Final Counts from Telescope:
OV_counts.telescope.tmp <- lapply(samples,
                           function(s){
                               tmp <- read.table(
                                   file.path('/Users/hchintalapudi/Desktop/work/TCGA analysis/PanCancer/TCGA-OV/Telescope',paste0(s, '_sorted-telescope_report.tsv')),
                                   sep='\t', header=T, stringsAsFactors=F)
                               ret <- data.frame(transcript=genes$transcript_id, stringsAsFactors=F) %>%
                                   left_join(tmp, by='transcript') %>%
                                   dplyr::mutate(
                                       gene_id = transcript,
                                       count = final_count
                                   ) %>%
                                   dplyr::select(gene_id, count)
                               ret[is.na(ret)] <- 0
                               stopifnot(all(ret$gene_id == genes$transcript_id))
                               ret$gene_id <- NULL
                               names(ret) <- c(s)
                               ret
                           }) %>%
    bind_cols
rownames(OV_counts.telescope)<- OV_counts.telescope.tmp$`00d4eec7-e2e3-4c52-958d-d181101fe0ec_gdc_realn_rehead`
table(OV_metadata$File.Name == colnames(OV_counts.telescope))
colnames(OV_counts.telescope)<- OV_metadata$Sample.ID
write.csv(OV_counts.telescope, file = "/Users/hchintalapudi/Desktop/work/TCGA analysis/PanCancer/TCGA-OV_Telescope_finalcounts.csv")
OV_counts.telescope<- read.csv("/Users/hchintalapudi/Desktop/work/TCGA analysis/PanCancer/TCGA-OV_Telescope_finalcounts.csv", stringsAsFactors = F, row.names = 1, check.names = F)
## subset to 365:
OV_counts.telescope_filt<- OV_counts.telescope[colnames(OV_counts.telescope) %in% ov_samples_filt_Ali$Sample.ID]
write.csv(OV_counts.telescope_filt, file = "/Users/hchintalapudi/Desktop/work/TCGA analysis/PanCancer/TCGA-OV_Telescope_finalcounts_filt.csv")
OV_counts.telescope_filt<- read.csv("/Users/hchintalapudi/Desktop/work/TCGA analysis/PanCancer/TCGA-OV_Telescope_finalcounts_filt.csv", stringsAsFactors = F, row.names = 1)
```


```{r}
## metadata cleanup:
a<- read.delim("/Users/hchintalapudi/Desktop/work/TCGA analysis/PanCancer/TCGA-OV/annotations.txt")
b_og<- read.delim("/Users/hchintalapudi/Desktop/work/TCGA analysis/PanCancer/TCGA-OV/nationwidechildrens.org_clinical_patient_ov.txt")
b_og<- b_og[-c(1,2),]
b_og<-b_og%>%mutate(bcr_patient_uuid = tolower(bcr_patient_uuid))
b$clinical_stage[b$clinical_stage == "[Not Available]"]<- NA
b$clinical_stage[b$clinical_stage == "Stage IA" |b$clinical_stage == "Stage IB" | b$clinical_stage == "Stage IC"]<- "Stage I"
b$clinical_stage[b$clinical_stage == "Stage IIA" |b$clinical_stage == "Stage IIB" | b$clinical_stage == "Stage IIC"]<- "Stage II"
b$clinical_stage[b$clinical_stage == "Stage IIIA" |b$clinical_stage == "Stage IIIB" | b$clinical_stage == "Stage IIIC"]<- "Stage III"

b$tumor_grade[b$tumor_grade == "[Not Available]"]<- NA

b<- b[,c("bcr_patient_barcode","age_at_initial_pathologic_diagnosis", "clinical_stage","vital_status", "tumor_grade", "last_contact_days_to","death_days_to")]
set.seed(12345)
b$survival.time <- as.integer(ifelse(b$death_days_to == "[Not Applicable]",
                                   b$last_contact_days_to, b$death_days_to))

b<- b[-c(6,7)]
b<- b%>%
  mutate(censor = ifelse(vital_status == "Alive", 0, ifelse(vital_status == "Dead", 1, NA)))


ov_surv_metadata<- left_join(OV_metadata[,c(6,7)], b[,-4], by = c("Case.ID" = "bcr_patient_barcode"))
write.csv(ov_surv_metadata, file = "/Users/hchintalapudi/Desktop/work/TCGA analysis/PanCancer/TCGA-OV/ov_surv_metadata.csv", row.names = F)

ov_surv_metadata<- read.csv("/Users/hchintalapudi/Desktop/work/TCGA analysis/PanCancer/TCGA-OV/ov_surv_metadata.csv", check.names = F, stringsAsFactors = F)

clinical<- read.delim("/Users/hchintalapudi/Desktop/work/TCGA analysis/PanCancer/TCGA-OV/clinical.cart.2021-09-02/clinical.tsv")

## new metadata file from cbioportal:
## http://www.cbioportal.org/study/clinicalData?id=ov_tcga_pan_can_atlas_2018
b.1<- read.delim("/Users/hchintalapudi/Desktop/work/TCGA analysis/PanCancer/TCGA-OV/ov_tcga_pan_can_atlas_2018_clinical_data.tsv")


## SIZE Factors for DESeq2:
metrics.list <- lapply(samples, 
                       function(s){
                           f <- file.path('/Users/hchintalapudi/Desktop/work/TCGA analysis/PanCancer/TCGA-OV/Telescope/',paste0(s, '_sorted-telescope_report.tsv'))
                           if(file.exists(f)) {
                               print(f)
                               h <- readLines(f, 1) %>% strsplit(., '\t') %>% unlist
                               rstr <- sapply(strsplit(h[-c(1,2)], ':'), function(t) as.numeric(unlist(t[2][1])))
                               names(rstr) <- sapply(strsplit(h[-c(1,2)], ':'), function(t) t[1])
                           } else{
                               rstr <- c(NA)
                           }
                           rstr
                       }
)
mn <- unique(do.call(c, lapply(metrics.list, names)))
met.ts <- lapply(metrics.list, 
                 function(m) {
                     ret <- sapply(mn, function(x) m[x])
                     names(ret) <- gsub('\\..*', '', names(ret))
                     ret
                 }) %>% do.call(rbind, .) %>% data.frame
names(met.ts) <- paste(names(met.ts), 'ts', sep='.')
met.ts$sample<- samples
ov_metrics<- met.ts
ov_metrics$sample
ov_metrics<- left_join(ov_metrics, OV_metadata[,c(1, 2,7)], by = c("sample"="File.Name"))


#--- Add library size (total frags mapped) columns
ov_metrics$mapped_frags.bt2 <- ov_metrics$pair_mapped.ts + ov_metrics$pair_mixed.ts + ov_metrics$single_mapped.ts
write.csv(ov_metrics, file = "/Users/hchintalapudi/Desktop/work/TCGA analysis/PanCancer/TCGA-OV/ov_metrics.csv")
# Function for calculating size factors
calculateSizeFactors <- function(mapped_frags) {
    geomean <- expm1(mean(log1p(mapped_frags)))
    mapped_frags / geomean
}
ov_sizefactors<- as.data.frame(calculateSizeFactors(ov_metrics$mapped_frags.bt2))
ov_sizefactors$Sample<- OV_metadata$Sample.ID
ov_sizefactors<- ov_sizefactors[,c(2,1)]
colnames(ov_sizefactors)<- c("Sample","sizefactor")
write.csv(ov_sizefactors, file = "/Users/hchintalapudi/Desktop/work/TCGA analysis/PanCancer/TCGA-OV/Telescope_ov_sizefactors.csv", row.names = F)
ov_sizefactors<- read.csv("/Users/hchintalapudi/Desktop/work/TCGA analysis/PanCancer/TCGA-OV/Telescope_ov_sizefactors.csv", row.names = 1, stringsAsFactors = F)

## get vst counts:
ov_surv_metadata<- ov_surv_metadata[complete.cases(ov_surv_metadata),]
ov_metadata_deseq<- ov_surv_metadata[,c(2,6)]
median(ov_metadata_deseq$survival.time)
rownames(ov_metadata_deseq)<- NULL
ov_metadata_deseq<- ov_metadata_deseq%>%
  mutate(Survival = ifelse(survival.time>=915, "LTS", ifelse(survival.time<=915, "STS",NA)))%>%
  column_to_rownames("Sample.ID")%>%
  dplyr::select(-1)

OV_counts.telescope_filt<- OV_counts.telescope[rownames(ov_metadata_deseq)]
table(colnames(OV_counts.telescope_filt) == rownames(ov_metadata_deseq))

dds_ov<- DESeqDataSetFromMatrix(countData = OV_counts.telescope_filt, colData =ov_metadata_deseq, design = ~Survival)
dds_ov$Survival<- factor(dds_ov$Survival, levels = c("STS", "LTS"))
ov_sizefactors_filt<- ov_sizefactors[ov_sizefactors$Sample %in% colnames(OV_counts.telescope_filt),]
ov_sizefactors<- ov_sizefactors[match(rownames(ov_metadata_deseq), ov_sizefactors$Sample),]
table(rownames(ov_metadata_deseq) == ov_sizefactors$Sample)

sizeFactors(dds_ov)<- ov_sizefactors$sizefactor
dds_ov.2<- dds_ov[rowSums(counts(dds_ov)>=10)>=2,]
table(rowSums(counts(dds_ov)>=10)>=2)

ov_vst<- as.data.frame(assay(vst(dds_ov.2, blind = F)))
write.csv(ov_vst, file = "/Users/hchintalapudi/Desktop/work/TCGA analysis/PanCancer/TCGA-OV/ov_vst_8355.csv")
ov_vst<- read.csv("/Users/hchintalapudi/Desktop/work/TCGA analysis/PanCancer/TCGA-OV/ov_vst_8355.csv", row.names = 1, check.names = F)

## filter ov_surv_metadata to contain 365 samples:
ov_surv_metadata<- ov_surv_metadata[ov_surv_metadata$Sample.ID %in% ov_samples_filt_Ali$Sample.ID,]
ov_vst<- ov_vst[colnames(ov_vst) %in% ov_samples_filt_Ali$Sample.ID]
table(colnames(ov_vst) == ov_surv_metadata$Sample.ID)

ov_cox_input<- data.frame(ov_surv_metadata[,c("Sample.ID","survival.time","censor","age_at_initial_pathologic_diagnosis", "clinical_stage","tumor_grade")], t(ov_vst))
rownames(ov_cox_input)<- NULL
ov_cox_input<- ov_cox_input%>%
  column_to_rownames("Sample.ID")

ov_cox_input$survival.time<- as.numeric(ov_cox_input$survival.time)
ov_cox_input$censor<- as.numeric(ov_cox_input$censor)
ov_cox_input$age_at_initial_pathologic_diagnosis<- as.numeric(ov_cox_input$age_at_initial_pathologic_diagnosis)
ov_cox_input[,c(4,5)]<- lapply(ov_cox_input[,c(4,5)], as.factor)


ov_samples_filt_Ali<- read.table("/Users/hchintalapudi/Desktop/work/TCGA analysis/PanCancer/TCGA-OV/case_set_OV_ERV__no_prior_tx.2021-11-18.tsv", header = T)





#ov_clinical<- read.delim("/Users/hchintalapudi/Desktop/work/TCGA analysis/PanCancer/TCGA-OV/clinical.cases_selection.2022-03-16 (1)/clinical.tsv")
ov_samples_filt_Ali<-left_join(ov_samples_filt_Ali, b_og[,c(1,2)], by = c("id" = "bcr_patient_uuid"))
ov_samples_filt_Ali<- left_join(ov_samples_filt_Ali, OV_metadata[,c(1,2,6,7)],  by = c("bcr_patient_barcode" = "Case.ID"))

## removing the 2 samples with uuids not matching sample ids as they're 02 samples**:
ov_samples_filt_Ali<- ov_samples_filt_Ali[complete.cases(ov_samples_filt_Ali),]

##'bcdf0550-9e6f-4500-9473-706df7beeae4_gdc_realn_rehead.bam' is a truncated bam file: bam2fq fails

write.csv(ov_samples_filt_Ali, file = "/Users/hchintalapudi/Desktop/work/TCGA analysis/PanCancer/TCGA-OV/case_set_OV_ERV__no_prior_tx.2021-11-18_updated.tsv")
ov_qual.cases<- read.table("/Users/hchintalapudi/Desktop/work/TCGA analysis/PanCancer/TCGA-OV/case_set_OV_ERV__no_prior_tx.2021-11-18_updated.tsv", sep = ",", header = T, row.names = 1)
```


```{r}
ov_cox_input<- ov_cox_input[rownames(ov_cox_input) %in% tmp$Sample.ID,]
write.csv(ov_cox_input, file = "/Users/hchintalapudi/Desktop/work/TCGA analysis/PanCancer/TCGA-OV/ov_cox_input.csv")
write.csv(ov_cox_input, file = "/Users/hchintalapudi/Desktop/work/TCGA analysis/PanCancer/TCGA-OV/ov_cox_input_filt.csv")
ov_cox_input<- read.csv("/Users/hchintalapudi/Desktop/work/TCGA analysis/PanCancer/TCGA-OV/ov_cox_input.csv", row.names = 1, check.names = F)

## Cox with RegParalell:
ov_cox_res<-RegParallel(
    data = ov_cox_input,
    formula = 'Surv(survival.time, censor) ~ [*] + age_at_initial_pathologic_diagnosis + clinical_stage + tumor_grade',
    FUN = function(formula, data)
      coxph(formula = formula,
        data = data,
        ties = 'breslow',
        singular.ok = TRUE),
    FUNtype = 'coxph',
    variables = colnames(ov_cox_input)[6:ncol(ov_cox_input)],
    blocksize = 1000,
    cores = 2,
    nestedParallel = FALSE,
    conflevel = 95,
    excludeTerms = c("age_at_initial_pathologic_diagnosis", "clinical_stage", "tumor_grade"),
    excludeIntercept = TRUE,
    p.adjust = 'fdr')
write.csv(ov_cox_res, file = "/Users/hchintalapudi/Desktop/work/TCGA analysis/PanCancer/TCGA-OV/OV-vst_cox_results.csv")
ov_cox_res<- read.csv("/Users/hchintalapudi/Desktop/work/TCGA analysis/PanCancer/TCGA-OV/OV-vst_cox_results.csv", row.names = 1)

ov_unstr<- read.delim("/Users/hchintalapudi/Desktop/work/TCGA analysis/PanCancer/TCGA-OV/TCGA-OV_unstr.txt", check.names = F, row.names = 1)
ov_unstr<- ov_unstr[samples]
table(samples == colnames(ov_unstr))
table(samples == OV_metadata$File.Name)
colnames(ov_unstr)<- OV_metadata$Sample.ID

write.csv(ov_unstr, file = "/Users/hchintalapudi/Desktop/work/TCGA analysis/PanCancer/TCGA-OV/TCGA-OV_unstr.2.txt")
ov_unstr<- read.table("/Users/hchintalapudi/Desktop/work/TCGA analysis/PanCancer/TCGA-OV/TCGA-OV_unstr.2.txt", check.names = F, row.names = 1, sep = ",", header = T)
# subset to 365:
ov_unstr_filt<- ov_unstr[colnames(ov_unstr) %in% ov_samples_filt_Ali$Sample.ID]
```


```{r}
## TPM:



table(rownames(ov_unstr) == gene_lengths$gene_name)
table(rownames(ov_unstr_filt) == gene_lengths$gene_name)

ov_genecounts_tpm<- as.data.frame(tpm3(ov_unstr, gene_lengths$gene_length))
write.csv(ov_genecounts_tpm, file = "/Users/hchintalapudi/Desktop/work/TCGA analysis/PanCancer/TCGA-OV/ov_genecounts_tpm.csv")

ov_genecounts_AM_tpm<- ov_genecounts_tpm[rownames(ov_cox_input)]
write.csv(ov_genecounts_AM_tpm, file = "/Users/hchintalapudi/Desktop/work/TCGA analysis/PanCancer/TCGA-OV/ov_genecounts_AM_tpm.csv")
ov_genecounts_AM_tpm<- read.csv("/Users/hchintalapudi/Desktop/work/TCGA analysis/PanCancer/TCGA-OV/ov_genecounts_AM_tpm.csv", row.names = 1, check.names = F)

## TPMs of 365:
ov_genecounts_AM_tpm_filt<- as.data.frame(tpm3(ov_unstr_filt, gene_lengths$gene_length))
table(colnames(ov_genecounts_AM_tpm_filt) == rownames(ov_cox_input))
write.csv(ov_genecounts_AM_tpm_filt, file = "/Users/hchintalapudi/Desktop/work/TCGA analysis/PanCancer/TCGA-OV/ov_genecounts_AM_tpm_filt.csv")
```


```{r}
# Calc CPM:
OV_counts.telescope.cpm<- OV_counts.telescope
table(colnames(OV_counts.telescope.cpm) == ov_sizefactors$Sample)
OV_counts.telescope.cpm <- (OV_counts.telescope/ov_metrics$mapped_frags.bt2)*1000000

write.csv(OV_counts.telescope.cpm, file = "/Users/hchintalapudi/Desktop/work/TCGA analysis/PanCancer/TCGA-OV/Telescope_ov_CPM.csv")
OV_counts.telescope.cpm<- read.csv("/Users/hchintalapudi/Desktop/work/TCGA analysis/PanCancer/TCGA-OV/Telescope_ov_CPM.csv", stringsAsFactors = F, check.names = F)

OV_counts.telescope.cpm.filt<- OV_counts.telescope.cpm[colnames(OV_counts.telescope.cpm) %in% rownames(ov_cox_input)]
write.csv(OV_counts.telescope.cpm.filt, file = "/Users/hchintalapudi/Desktop/work/TCGA analysis/PanCancer/TCGA-OV/Telescope_ov_CPM_filt.csv")
table((rowSums(OV_counts.telescope.cpm.filt)>0.2)>=length(colnames(OV_counts.telescope.cpm.filt))*0.75)

```



