---
title: "TCGA-LIHC"
author: "Himanshu Chintalapudi"
date: "5/10/2022"
output: html_document
---

```{r}
source("libs.R")
## TCGA-LIHC:
genes<- as.data.frame(rtracklayer::import("/Users/hchintalapudi/Desktop/work/TCGA analysis/genes.gtf"))

LIHC_metadata<- read.table("/Users/hchintalapudi/Desktop/work/TCGA analysis/PanCancer/TCGA-LIHC/LIHC_gdc_sample_sheet.2021-08-12.tsv", header = T, sep = "\t")
LIHC_metadata<- LIHC_metadata[LIHC_metadata$Sample.Type == "Primary Tumor",]

LIHC_metadata$File.Name<- gsub(x = LIHC_metadata$File.Name, pattern = ".bam", replacement = "")
samples<- as.character(LIHC_metadata$File.Name)
#--- Final Counts from Telescope:
LIHC_counts.telescope <- lapply(samples,
                           function(s){
                               tmp <- read.table(
                                   file.path('/Users/hchintalapudi/Desktop/work/TCGA analysis/PanCancer/TCGA-LIHC/Telescope',paste0(s, '_sorted-telescope_report.tsv')),
                                   sep='\t', header=T, stringsAsFactors=F)
                               ret <- data.frame(transcript=genes$transcript_id, stringsAsFactors=F) %>%
                                   left_join(tmp, by='transcript') %>%
                                   dplyr::mutate(
                                       gene_id = transcript,
                                       count = final_count
                                   ) %>%
                                   dplyr::select(gene_id, count)
                               ret[is.na(ret)] <- 0
                               stopifnot(all(ret$gene_id == genes$transcript_id))
                               ret$gene_id <- NULL
                               names(ret) <- c(s)
                               ret
                           }) %>%
    bind_cols
rownames(LIHC_counts.telescope)<- LIHC_counts.telescope.tmp$`70135ee9-90ab-4ecb-9723-99de6c161573_gdc_realn_rehead`
table(LIHC_metadata$File.Name == colnames(LIHC_counts.telescope))
colnames(LIHC_counts.telescope)<- LIHC_metadata$Sample.ID
write.csv(LIHC_counts.telescope, file = "/Users/hchintalapudi/Desktop/work/TCGA analysis/PanCancer/TCGA-LIHC_Telescope_finalcounts.csv")
LIHC_counts.telescope<- read.csv("/Users/hchintalapudi/Desktop/work/TCGA analysis/PanCancer/TCGA-LIHC_Telescope_finalcounts.csv", row.names = 1, check.names = F)

## metadata cleanup:
l1<- read.delim("/Users/hchintalapudi/Desktop/work/TCGA analysis/PanCancer/TCGA-LIHC/annotations.txt")
l2<- read.delim("/Users/hchintalapudi/Desktop/work/TCGA analysis/PanCancer/TCGA-LIHC/nationwidechildrens.org_clinical_patient_lihc.txt")
l2<- l2[-c(1,2),]
#l3<- read.delim("/Users/hchintalapudi/Desktop/work/TCGA analysis/PanCancer/TCGA-LIHC/")

l2<- l2[,c("bcr_patient_barcode","vital_status","ajcc_pathologic_tumor_stage", "tumor_grade","age_at_diagnosis",  "last_contact_days_to","death_days_to")]
l2$age_at_diagnosis[l2$age_at_diagnosis == "[Not Available]"]<- NA
l2<- left_join(l2, LIHC_metadata[,c(6,7)], by = c("bcr_patient_barcode" = "Case.ID"))

l2$ajcc_pathologic_tumor_stage[l2$ajcc_pathologic_tumor_stage == "[Discrepancy]" | l2$ajcc_pathologic_tumor_stage == "[Not Available]"]<- NA
l2$ajcc_pathologic_tumor_stage[l2$ajcc_pathologic_tumor_stage == "Stage IIIA" | l2$ajcc_pathologic_tumor_stage == "Stage IIIB" | l2$ajcc_pathologic_tumor_stage == "Stage IIIC"]<- "Stage III"

l2$ajcc_pathologic_tumor_stage[l2$ajcc_pathologic_tumor_stage == "Stage IVA" | l2$ajcc_pathologic_tumor_stage == "Stage IVB"]<- "Stage IV"

l2$tumor_grade[l2$tumor_grade == "[Not Available]"]<- NA

l2<- l2%>%
  mutate(censor = ifelse(vital_status == "Alive", 0, ifelse(vital_status == "Dead", 1, NA)))

l2$survival.time <- as.integer(ifelse(l2$death_days_to == "[Not Applicable]",
                                   l2$last_contact_days_to, l2$death_days_to))
l2$survival.time[l2$survival.time<0]<- 0
l2.1<- l2[,c(8,9,10,5,3,4)]
l2.1<- l2.1[complete.cases(l2.1),]
l2.1<- l2.1[l2.1$Sample.ID %in% lihc_samples_filt_Ali$Sample.ID,]
write.csv(l2.1, file = "/Users/hchintalapudi/Desktop/work/TCGA analysis/PanCancer/TCGA-LIHC/lihc_surv_metadata.csv", row.names = F)
l2.1<- read.csv("/Users/hchintalapudi/Desktop/work/TCGA analysis/PanCancer/TCGA-LIHC/lihc_surv_metadata.csv")


LIHC_counts.telescope_filt<- LIHC_counts.telescope[colnames(LIHC_counts.telescope) %in% l2.1$Sample.ID]
lihc_metadata_deseq<- l2.1[,c(1,3)]
median(lihc_metadata_deseq$survival.time)

## SIZE Factors for DESeq2:
metrics.list <- lapply(samples, 
                       function(s){
                           f <- file.path('/Users/hchintalapudi/Desktop/work/TCGA analysis/PanCancer/TCGA-LIHC/Telescope/',paste0(s, '_sorted-telescope_report.tsv'))
                           if(file.exists(f)) {
                               print(f)
                               h <- readLines(f, 1) %>% strsplit(., '\t') %>% unlist
                               rstr <- sapply(strsplit(h[-c(1,2)], ':'), function(t) as.numeric(unlist(t[2][1])))
                               names(rstr) <- sapply(strsplit(h[-c(1,2)], ':'), function(t) t[1])
                           } else{
                               rstr <- c(NA)
                           }
                           rstr
                       }
)
mn <- unique(do.call(c, lapply(metrics.list, names)))
met.ts <- lapply(metrics.list, 
                 function(m) {
                     ret <- sapply(mn, function(x) m[x])
                     names(ret) <- gsub('\\..*', '', names(ret))
                     ret
                 }) %>% do.call(rbind, .) %>% data.frame
names(met.ts) <- paste(names(met.ts), 'ts', sep='.')
met.ts$sample<- samples
lihc_metrics<- met.ts
lihc_metrics$sample
lihc_metrics<- left_join(lihc_metrics, LIHC_metadata[,c(1, 2,6,7)], by = c("sample"="File.Name"))


#--- Add library size (total frags mapped) columns
lihc_metrics$mapped_frags.bt2 <- lihc_metrics$pair_mapped.ts + lihc_metrics$pair_mixed.ts + lihc_metrics$single_mapped.ts
write.csv(lihc_metrics, file = "/Users/hchintalapudi/Desktop/work/TCGA analysis/PanCancer/TCGA-LIHC/Telescope_lihc_metrics.csv")
lihc_metrics<- read.csv("/Users/hchintalapudi/Desktop/work/TCGA analysis/PanCancer/TCGA-LIHC/Telescope_lihc_metrics.csv", row.names = 1, stringsAsFactors = F)

# Function for calculating size factors
calculateSizeFactors <- function(mapped_frags) {
    geomean <- expm1(mean(log1p(mapped_frags)))
    mapped_frags / geomean
}
lihc_sizefactors<- as.data.frame(calculateSizeFactors(lihc_metrics$mapped_frags.bt2))
lihc_sizefactors$Sample<- LIHC_metadata$Sample.ID
lihc_sizefactors<- lihc_sizefactors[,c(2,1)]
colnames(lihc_sizefactors)<- c("Sample","sizefactor")
write.csv(lihc_sizefactors, file = "/Users/hchintalapudi/Desktop/work/TCGA analysis/PanCancer/TCGA-LIHC/Telescope_lihc_sizefactors.csv", row.names = F)
lihc_sizefactors<- read.csv("/Users/hchintalapudi/Desktop/work/TCGA analysis/PanCancer/TCGA-LIHC/Telescope_lihc_sizefactors.csv")

## get vst counts:

lihc_metadata_deseq<- l2.1[,c(1,3)]
median(lihc_metadata_deseq$survival.time)
rownames(lihc_metadata_deseq)<- NULL
lihc_metadata_deseq<- lihc_metadata_deseq%>%
  mutate(Survival = ifelse(survival.time>=344.5, "LTS", ifelse(survival.time<=344.5, "STS",NA)))%>%
  column_to_rownames("Sample.ID")%>%
  dplyr::select(-1)
write.csv(lihc_metadata_deseq, file = "/Users/hchintalapudi/Desktop/work/TCGA analysis/PanCancer/TCGA-LIHC/LIHC_metadata_deseq2.csv")
lihc_metadata_deseq<- read.csv("/Users/hchintalapudi/Desktop/work/TCGA analysis/PanCancer/TCGA-LIHC/LIHC_metadata_deseq2.csv", row.names = 1)

LIHC_counts.telescope_filt<- LIHC_counts.telescope[rownames(lihc_metadata_deseq)]
table(colnames(LIHC_counts.telescope_filt) == rownames(lihc_metadata_deseq))

dds_lihc<- DESeqDataSetFromMatrix(countData = LIHC_counts.telescope_filt, colData =lihc_metadata_deseq, design = ~Survival)
dds_lihc$Survival<- factor(dds_lihc$Survival, levels = c("STS", "LTS"))
lihc_sizefactors_filt<- lihc_sizefactors[lihc_sizefactors$Sample %in% colnames(LIHC_counts.telescope_filt),]
lihc_sizefactors_filt<- lihc_sizefactors_filt[match(rownames(lihc_metadata_deseq), lihc_sizefactors_filt$Sample),]
table(rownames(lihc_metadata_deseq) == lihc_sizefactors_filt$Sample)

sizeFactors(dds_lihc)<- lihc_sizefactors_filt$sizefactor
## minimum of 10 counts in 50% of patients:
dds_lihc.2<- dds_lihc[rowSums(counts(dds_lihc)>=10)>=171,]
table(rowSums(counts(dds_lihc)>=10)>= 171)

lihc_vst<- as.data.frame(assay(vst(dds_lihc.2, blind = F)))
write.csv(lihc_vst, file = "/Users/hchintalapudi/Desktop/work/TCGA analysis/PanCancer/TCGA-LIHC/lihc_vst_0.50_2229.csv")
lihc_vst<- read.csv("/Users/hchintalapudi/Desktop/work/TCGA analysis/PanCancer/TCGA-LIHC/lihc_vst_0.50_2229.csv", row.names = 1, check.names = F)




table(colnames(lihc_vst) == l2.1$Sample.ID)

lihc_cox_input<- data.frame(l2.1[,c("Sample.ID","survival.time","censor","age_at_diagnosis", "ajcc_pathologic_tumor_stage","tumor_grade")], t(lihc_vst))
rownames(lihc_cox_input)<- NULL
lihc_cox_input<- lihc_cox_input%>%
  column_to_rownames("Sample.ID")

lihc_cox_input$survival.time<- as.numeric(lihc_cox_input$survival.time)
lihc_cox_input$censor<- as.numeric(lihc_cox_input$censor)
lihc_cox_input$age_at_diagnosis<- as.numeric(lihc_cox_input$age_at_diagnosis)
lihc_cox_input[,c(4,5)]<- lapply(lihc_cox_input[,c(4,5)], as.factor)

#* 02/21/2022: (after filtering to 75% threshold as coded above)
#write.csv(lihc_cox_input, file = "/Users/hchintalapudi/Desktop/work/TCGA analysis/PanCancer/TCGA-LIHC/lihc_cox_input_0.75_1558.csv")
#* 02/28/2022: (after filtering to 50% threshold as coded above)
write.csv(lihc_cox_input, file = "/Users/hchintalapudi/Desktop/work/TCGA analysis/PanCancer/TCGA-LIHC/lihc_cox_input_0.50_2229.csv")




lihc_samples_filt_Ali<- read.table("/Users/hchintalapudi/Desktop/work/TCGA analysis/PanCancer/TCGA-LIHC/case_set_LIHC_ERV__no_prior_tx.2021-11-18.tsv", header = T)
tmp<- read.table("/Users/hchintalapudi/Desktop/work/TCGA analysis/PanCancer/TCGA-LIHC/case_set_LIHC_ERV__no_prior_tx.2021-11-18.tsv", header = T)
l2<-l2%>%mutate(bcr_patient_uuid = tolower(bcr_patient_uuid))
lihc_samples_filt_Ali<- left_join(lihc_samples_filt_Ali, l2[,c(1,2)], by = c("id" = "bcr_patient_uuid"))
lihc_samples_filt_Ali<- left_join(lihc_samples_filt_Ali, LIHC_metadata[,c(7,6)],  by = c("bcr_patient_barcode" = "Case.ID"))
lihc_samples_filt_Ali<- lihc_samples_filt_Ali[!grepl("-11A|-11B|-02", lihc_samples_filt_Ali$Sample.ID),]
write.csv(lihc_samples_filt_Ali, file = "/Users/hchintalapudi/Desktop/work/TCGA analysis/PanCancer/TCGA-LIHC/case_set_LIHC_ERV__no_prior_tx.2021-11-18_updated.tsv")

# subset to 342:
lihc_cox_input<- lihc_cox_input[rownames(lihc_cox_input) %in% lihc_samples_filt_Ali$Sample.ID,]
write.csv(lihc_cox_input, file = "/Users/hchintalapudi/Desktop/work/TCGA analysis/PanCancer/TCGA-LIHC/lihc_cox_input_filt.csv")

## Cox with RegParalell:
lihc_cox_res<-RegParallel(
    data = lihc_cox_input,
    formula = 'Surv(survival.time, censor) ~ [*] + age_at_diagnosis + ajcc_pathologic_tumor_stage + tumor_grade',
    FUN = function(formula, data)
      coxph(formula = formula,
        data = data,
        ties = 'breslow',
        singular.ok = TRUE),
    FUNtype = 'coxph',
    variables = colnames(lihc_cox_input)[6:ncol(lihc_cox_input)],
    blocksize = 1000,
    cores = 2,
    nestedParallel = FALSE,
    conflevel = 95,
    excludeTerms = c("age_at_diagnosis", "ajcc_pathologic_tumor_stage", "tumor_grade"),
    excludeIntercept = TRUE,
    p.adjust = 'fdr')
write.csv(lihc_cox_res, file = "/Users/hchintalapudi/Desktop/work/TCGA analysis/PanCancer/TCGA-LIHC/LIHC-vst_cox_results_0.50_2229.csv")

lihc_cox_res<- read.csv("/Users/hchintalapudi/Desktop/work/TCGA analysis/PanCancer/TCGA-LIHC/LIHC-vst_cox_results_0.50_2229.csv", row.names = 1)

lihc_unstr<- read.delim("/Users/hchintalapudi/Desktop/work/TCGA analysis/PanCancer/TCGA-LIHC/TCGA-LIHC_unstr-counts.txt", check.names = F, row.names = 1, stringsAsFactors = F)
lihc_unstr<- lihc_unstr[samples]
table(samples == colnames(lihc_unstr))
table(samples == LIHC_metadata$File.Name)
colnames(lihc_unstr)<- LIHC_metadata$Sample.ID


write.csv(lihc_unstr, file = "/Users/hchintalapudi/Desktop/work/TCGA analysis/PanCancer/TCGA-LIHC/TCGA-LIHC_unstr.2.txt")
lihc_genecounts<- read.csv("/Users/hchintalapudi/Desktop/work/TCGA analysis/PanCancer/TCGA-LIHC/TCGA-LIHC_unstr.2.txt", row.names = 1, check.names = F, stringsAsFactors = F)
lihc_genecounts.filt<- lihc_genecounts[rownames(lihc_cox_input)]
write.csv(lihc_genecounts.filt, file = "/Users/hchintalapudi/Desktop/work/TCGA analysis/PanCancer/TCGA-LIHC/TCGA-LIHC_genecounts_unstr_raw_342.txt")

##--------- Calc. vst for WGCNA analysis (genes + ERVs):
table(colnames(lihc_genecounts.filt) == rownames(lihc_metadata_deseq))
dds_lihc.g<- DESeqDataSetFromMatrix(countData = lihc_genecounts.filt, colData =lihc_metadata_deseq, design = ~Survival)
dds_lihc.g$Survival<- factor(dds_lihc.g$Survival, levels = c("STS", "LTS"))
lihc_sizefactors_filt<- lihc_sizefactors[lihc_sizefactors$Sample %in% colnames(lihc_genecounts.filt),]
lihc_sizefactors_filt<- lihc_sizefactors_filt[match(rownames(lihc_metadata_deseq), lihc_sizefactors_filt$Sample),]
table(rownames(lihc_metadata_deseq) == lihc_sizefactors_filt$Sample)
sizeFactors(dds_lihc.g)<- lihc_sizefactors_filt$sizefactor

## minimum of 10 counts in 50% of patients:
dds_lihc.g.2<- dds_lihc.g[rowSums(counts(dds_lihc.g)>=10)>=170,]
table(rowSums(counts(dds_lihc.g)>=10)>= 170)

lihc_genecounts.vst<- as.data.frame(assay(vst(dds_lihc.g.2, blind = F)))
write.csv(lihc_genecounts.vst, file = "/Users/hchintalapudi/Desktop/work/TCGA analysis/PanCancer/TCGA-LIHC/lihc_genecounts.vst_0.50_2229.csv")
##----------



#TPM:
table(rownames(lihc_unstr) == gene_lengths$gene_name)


lihc_genecounts_tpm<- as.data.frame(tpm3(lihc_unstr, gene_lengths$gene_length))
write.csv(lihc_genecounts_tpm, file = "/Users/hchintalapudi/Desktop/work/TCGA analysis/PanCancer/TCGA-LIHC/lihc_genecounts_AM_tpm_filt.csv")
lihc_genecounts_tpm<- read.csv("/Users/hchintalapudi/Desktop/work/TCGA analysis/PanCancer/TCGA-LIHC/lihc_genecounts_AM_tpm_filt.csv")

#02/02/2022:
# Calc CPM:
LIHC_counts.telescope.cpm<- LIHC_counts.telescope
table(colnames(LIHC_counts.telescope.cpm) == lihc_sizefactors$Sample)
LIHC_counts.telescope.cpm <- LIHC_counts.telescope.cpm/lihc_metrics$mapped_frags.bt2
LIHC_counts.telescope.cpm<- LIHC_counts.telescope.cpm*1000000

tmp<- (LIHC_counts.telescope/lihc_metrics$mapped_frags.bt2)*1000000

write.csv(LIHC_counts.telescope.cpm, file = "/Users/hchintalapudi/Desktop/work/TCGA analysis/PanCancer/TCGA-LIHC/Telescope_lihc_CPM.csv")
LIHC_counts.telescope.cpm<- read.csv("/Users/hchintalapudi/Desktop/work/TCGA analysis/PanCancer/TCGA-LIHC/Telescope_lihc_CPM.csv", stringsAsFactors = F, check.names = F, row.names = 1)

LIHC_counts.telescope.cpm.filt<- LIHC_counts.telescope.cpm[colnames(LIHC_counts.telescope.cpm) %in% rownames(lihc_cox_input)]

#Remove ERVs if count is > 0.2 in 75% of sample
rem <- function(x){
  x <- as.matrix(x)
  x <- t(apply(x,1,as.numeric))
  r <- as.numeric(apply(x,1,function(i) sum(i >= 0.2) ))
  remove <- which(r > dim(x)[2]*0.75)
  return(remove)
}
remove <- rem(LIHC_counts.telescope.cpm.filt)
LIHC_counts.telescope.cpm.filt.0.75 <- LIHC_counts.telescope.cpm.filt[-remove,]


table((rowSums(LIHC_counts.telescope.cpm.filt>=0.2)>=258))

var <- apply(LIHC_counts.telescope.cpm.filt, 1, var)
mean <- apply(LIHC_counts.telescope.cpm.filt, 1, mean)

plot(density(mock_var))
cor(mock_mean, mock_var, method="spearman")

plot(log2(mean), log2(var), pch='.')

plot(density(mean))
plot(log2(var))

df_meanVar <- data.frame(mean, var)

ggplot(df_meanVar) +
  geom_point(aes(x=mean, y= var)) +
  scale_y_log10() +
  scale_x_log10() +
  xlab("Mean counts per gene") +
  ylab("Variance per gene")

getVar <- apply(df, 1, var)
getVar.log<- log2(getVar)
param <- 0
## getting the top genes (CPM) by high variance, threshold on log2 scale:
df.2<- df[getVar.log >= param & !is.na(getVar.log), ] 

ggplot(df.2) +
  geom_point(aes(x=mean, y= var)) +
  scale_y_log10() +
  scale_x_log10() +
  xlab("Mean counts per gene") +
  ylab("Variance per gene")

LIHC_counts.telescope.filt<- LIHC_counts.telescope[rownames(lihc_cox_input)]

rv <- data.frame(matrixStats::rowVars(as.matrix((LIHC_counts.telescope.filt))), rownames(LIHC_counts.telescope.filt))
colnames(rv)<- c("rowVars", "ERV")
rv2 <- data.frame(Seq = seq(1:nrow(LIHC_counts.telescope.filt)), rowVars = rv[order(rv$rowVars, decreasing = TRUE),])
theme_set(theme_bw(base_size = 10))
ggplot(rv2, aes(x=Seq,y=rowVars.rowVars)) + geom_line() + scale_y_log10() +
  ggtitle("Telescope raw counts ordered by rowVar")

rv3<- rv2[1:5000,]
int.2<- intersect(rv3$rowVars.ERV, rownames(lihc_vst))

LIHC_CPM.topVar<- LIHC_counts.telescope.cpm.filt[rownames(LIHC_counts.telescope.cpm.filt) %in% rownames(df.2),]
LIHC_CPM.topVar<- as.data.frame(t(LIHC_CPM.topVar[rownames(lihc_cox_input)]))
## Trying Cox on the 116 ERVs & CPM counts instead of vst*

lihc_cox_input_CPM.topVar<- cbind(lihc_cox_input[,c(1:5)], LIHC_CPM.topVar)
lihc_cox_input_CPM.topVar[,c(1:3)]<- lapply(lihc_cox_input_CPM.topVar[,c(1:3)], as.numeric)
lihc_cox_input_CPM.topVar[,c(4,5)]<- lapply(lihc_cox_input_CPM.topVar[,c(4,5)], as.factor)

colnames(lihc_cox_input_CPM.topVar)[118]<- "ERV316A3_6_GL000256v2_alt.o"
lihc_cox_CPM.topVar_res<-RegParallel(
    data = lihc_cox_input_CPM.topVar,
    formula = 'Surv(survival.time, censor) ~ [*] + age_at_diagnosis + ajcc_pathologic_tumor_stage + tumor_grade',
    FUN = function(formula, data)
      coxph(formula = formula,
        data = data,
        ties = 'breslow',
        singular.ok = TRUE),
    FUNtype = 'coxph',
    variables = colnames(lihc_cox_input_CPM.topVar)[6:ncol(lihc_cox_input_CPM.topVar)],
    blocksize = 100,
    cores = 2,
    nestedParallel = FALSE,
    conflevel = 95,
    excludeTerms = c("age_at_diagnosis", "ajcc_pathologic_tumor_stage", "tumor_grade"),
    excludeIntercept = TRUE,
    p.adjust = 'fdr')

write.csv(lihc_cox_CPM.topVar_res, file = "/Users/hchintalapudi/Desktop/work/TCGA analysis/PanCancer/TCGA-LIHC/lihc_cox_results_CPM.topVar.csv")
write.csv(lihc_cox_input_CPM.topVar, file = "/Users/hchintalapudi/Desktop/work/TCGA analysis/PanCancer/TCGA-LIHC/lihc_cox_input_CPM.topVar.csv")
```


```{r}
## Sum up counts across ERV family level:
lihc_vst.tmp.tmp<- lihc_vst%>%
  rownames_to_column()%>%
  left_join(Telescope_locus.fam, by = c("rowname" = "transcript_id"))%>%
  column_to_rownames("rowname")

lihc_erv_fam_vst<- aggregate(lihc_vst.tmp.tmp[,c(1:342)], by = list(repFamily = lihc_vst.tmp.tmp$V1), FUN = sum)
lihc_erv_fam_vst<- lihc_erv_fam_vst%>%
  column_to_rownames("repFamily")
write.csv(lihc_erv_fam_vst, file = "/Users/hchintalapudi/Desktop/work/TCGA analysis/PanCancer/TCGA-LIHC/lihc_vst_60fam.csv")
lihc_erv_fam_vst<- read.csv("/Users/hchintalapudi/Desktop/work/TCGA analysis/PanCancer/TCGA-LIHC/lihc_vst_60fam.csv", row.names = 1, check.names = F)

lihc_cox_input_60fam<- cbind(lihc_cox_input[,c(1:5)], as.data.frame(t(lihc_erv_fam_vst)))
lihc_cox_res_60fam<-RegParallel(
    data = lihc_cox_input_60fam,
    formula = 'Surv(survival.time, censor) ~ [*] + age_at_diagnosis + ajcc_pathologic_tumor_stage + tumor_grade',
    FUN = function(formula, data)
      coxph(formula = formula,
        data = data,
        ties = 'breslow',
        singular.ok = TRUE),
    FUNtype = 'coxph',
    variables = colnames(lihc_cox_input_60fam)[6:ncol(lihc_cox_input_60fam)],
    blocksize = 50,
    cores = 2,
    nestedParallel = FALSE,
    conflevel = 95,
    excludeTerms = c("age_at_diagnosis", "ajcc_pathologic_tumor_stage", "tumor_grade"),
    excludeIntercept = TRUE,
    p.adjust = 'fdr')
write.csv()


```

