---
title: "TCGA-BRCA"
author: "Himanshu Chintalapudi"
date: "5/11/2022"
output: html_document
---

```{r}
source("libs.R")
```

```{r}
## TCGA-BRCA:
brca_manifest<- read.table("/Users/hchintalapudi/Desktop/work/TCGA analysis/PanCancer/TCGA-BRCA/gdc_manifest.2022-02-17.txt", header = T)
brca_manifest<- brca_manifest%>%
  mutate(id = tolower(id))



brca_samplesheet<- read.delim("/Users/hchintalapudi/Desktop/work/TCGA analysis/PanCancer/TCGA-BRCA/gdc_sample_sheet.2022-02-17.tsv")



brca_clinical<- read.delim("/Users/hchintalapudi/Desktop/work/TCGA analysis/PanCancer/TCGA-BRCA/nationwidechildrens.org_clinical_patient_brca.txt")
brca_clinical<- brca_tmp[-c(1,2),]

brca_clinical<- brca_clinical%>%
  mutate(bcr_patient_uuid = tolower(bcr_patient_uuid))

brca_q.cases<-read.table("/Users/hchintalapudi/Desktop/work/TCGA analysis/PanCancer/TCGA-BRCA/case_set_BRCA_ERV_no_tx_ductal_primary.2021-12-13.tsv", header = T)
brca_q.cases<- brca_q.cases%>%
  mutate(id = tolower(id))

brca_q.cases<- left_join(brca_q.cases, brca_clinical[,c(1,2)], by = c("id" = "bcr_patient_uuid"))
brca_q.cases<- left_join(brca_q.cases, brca_samplesheet[,c(1,2,6,7)], by = c("bcr_patient_barcode" = "Case.ID"))

brca_q.cases<- brca_q.cases[!grepl("-11A|-02|-11B", brca_q.cases$Sample.ID),]
rownames(brca_q.cases)<- NULL
brca_q.cases<- brca_q.cases[-258,]
## Removing "a06ff6ae-b0dd-4516-a75c-0db71694b46b" as no associated bam file or sample ID was found!
brca_q.cases<- brca_q.cases[!grepl("a06ff6ae-b0dd-4516-a75c-0db71694b46b", brca_q.cases$id),]

#check all the NA occurences:
brca_q.cases[is.na(brca_q.cases$Sample.ID),]
## bam files found in unedited samplesheet for all the NA occurences after left_join: 
##store these:
brca_discrepancies<-data.frame(brca_samplesheet[brca_samplesheet$File.Name %in%  c("44bb0b80-ac9b-4495-a014-a54eb389cf67_gdc_realn_rehead.bam",                                                       "dce39f40-597b-40df-961e-4d0b4f1e4a4a_gdc_realn_rehead.bam",  "eab0f8df-e01f-423f-bc98-dac806d56e13_gdc_realn_rehead.bam"),])
## MAnually add the case.ids to these discrepancies in the same order:
brca_discrepancies$id<- c("c31900a4-5dcd-4022-97ac-638e86e889e4", "427d0648-3f77-4ffc-b52c-89855426d647","6e7d5ec6-a469-467c-b748-237353c23416")
## change order of columns or subset to reflect original 'brca_q.cases' file:
brca_discrepancies.2<- brca_discrepancies[,c(9,6,1,2,7)]

## remove NAs and manually rbind the missing cases:
brca_q.cases<- brca_q.cases[complete.cases(brca_q.cases),]
colnames(brca_discrepancies.2)[2]<- "bcr_patient_barcode"
brca_q.cases<- rbind(brca_q.cases, brca_discrepancies.2)
## remove 1f6ab8ee-9f1d-4c71-b7ca-e77a922380c4_gdc_realn_rehead.bam & 1844cc21-f9ae-49f8-b383-299da0970bdd_gdc_realn_rehead.bam for duplicate occurence of id:6cb6f179-defd-4661-af0a-c353b74c0c49 as '12f301fc-bde2-4343-871e-acb4162043d7_gdc_realn_rehead.bam' has bigger read length; So retain only that
brca_q.cases<- brca_q.cases[!grepl('1f6ab8ee-9f1d-4c71-b7ca-e77a922380c4_gdc_realn_rehead.bam|1844cc21-f9ae-49f8-b383-299da0970bdd_gdc_realn_rehead.bam', brca_q.cases$File.Name),]
## remove '108dc6d1-5612-4276-a682-0ba4d324cb00_gdc_realn_rehead.bam' as it is also truncated: bam2fq fails:
brca_q.cases<- brca_q.cases[!brca_q.cases$File.Name %in% "108dc6d1-5612-4276-a682-0ba4d324cb00_gdc_realn_rehead.bam",]


write.csv(brca_q.cases, file = "/Users/hchintalapudi/Desktop/work/TCGA analysis/PanCancer/TCGA-BRCA/BRCA_qualified.cases.txt")
brca_qual_cases<- read.csv("/Users/hchintalapudi/Desktop/work/TCGA analysis/PanCancer/TCGA-BRCA/BRCA_qualified.cases.txt", row.names = 1)


#brca_manifest.q<- brca_manifest[brca_manifest$filename %in% brca_q.cases$File.Name,]
#rownames(brca_manifest.q)<- NULL
#write.table(brca_manifest.q, file = "/Users/hchintalapudi/Desktop/work/TCGA analysis/PanCancer/TCGA-BRCA/BRCA_manifest_q.cases.txt", quote = F, sep = "\t", row.names = F)
#brca_manifest.q<- read.table("/Users/hchintalapudi/Desktop/work/TCGA analysis/PanCancer/TCGA-BRCA/BRCA_manifest_q.cases.txt", header = T)


brca_qual_cases$File.Name<- gsub(x = brca_qual_cases$File.Name, pattern = ".bam", replacement = "")
samples<- as.character(brca_qual_cases$File.Name)

write.csv(samples, file = "/Users/hchintalapudi/Desktop/work/TCGA analysis/PanCancer/TCGA-BRCA/brca_q_samplenames.txt", row.names = F, quote = F)


files<- list.files(path='/Users/hchintalapudi/Desktop/work/TCGA analysis/PanCancer/TCGA-BRCA/Telescope/')
files<- gsub(x=as.list(files), pattern ='_sorted-telescope_report.tsv', replacement = "")

## Telescope:
#--- Final Counts from Telescope:
brca_counts.telescope <- lapply(samples,
                           function(s){
                               tmp <- read.table(
                                   file.path('/Users/hchintalapudi/Desktop/work/TCGA analysis/PanCancer/TCGA-BRCA/Telescope/',paste0(s, '_sorted-telescope_report.tsv')),
                                   sep='\t', header=T, stringsAsFactors=F)
                               ret <- data.frame(transcript=genes$transcript_id, stringsAsFactors=F) %>%
                                   left_join(tmp, by='transcript') %>%
                                   dplyr::mutate(
                                       gene_id = transcript,
                                       count = final_count
                                   ) %>%
                                   dplyr::select(gene_id, count)
                               ret[is.na(ret)] <- 0
                               stopifnot(all(ret$gene_id == genes$transcript_id))
                               ret$gene_id <- NULL
                               names(ret) <- c(s)
                               ret
                           }) %>%
    bind_cols
rownames(brca_counts.telescope)<- genes$transcript_id
brca_counts.telescope<- brca_counts.telescope[brca_qual_cases$File.Name]
table(brca_qual_cases$File.Name == colnames(brca_counts.telescope))
colnames(brca_counts.telescope)<- brca_qual_cases$Sample.ID

write.csv(brca_counts.telescope, file = "/Users/hchintalapudi/Desktop/work/TCGA analysis/PanCancer/TCGA-BRCA/TCGA-BRCA_Telescope_finalcounts_445.csv")


## SIZE Factors for DESeq2:
metrics.list <- lapply(samples, 
                       function(s){
                           f <- file.path('/Users/hchintalapudi/Desktop/work/TCGA analysis/PanCancer/TCGA-BRCA/Telescope/',paste0(s, '_sorted-telescope_report.tsv'))
                           if(file.exists(f)) {
                               print(f)
                               h <- readLines(f, 1) %>% strsplit(., '\t') %>% unlist
                               rstr <- sapply(strsplit(h[-c(1,2)], ':'), function(t) as.numeric(unlist(t[2][1])))
                               names(rstr) <- sapply(strsplit(h[-c(1,2)], ':'), function(t) t[1])
                           } else{
                               rstr <- c(NA)
                           }
                           rstr
                       }
)
mn <- unique(do.call(c, lapply(metrics.list, names)))
met.ts <- lapply(metrics.list, 
                 function(m) {
                     ret <- sapply(mn, function(x) m[x])
                     names(ret) <- gsub('\\..*', '', names(ret))
                     ret
                 }) %>% do.call(rbind, .) %>% data.frame
names(met.ts) <- paste(names(met.ts), 'ts', sep='.')
met.ts$sample<- samples
brca_metrics<- met.ts
brca_metrics$sample
brca_metrics<- left_join(brca_metrics, brca_qual_cases, by = c("sample"="File.Name"))


#--- Add library size (total frags mapped) columns
brca_metrics$mapped_frags.bt2 <- brca_metrics$pair_mapped.ts + brca_metrics$pair_mixed.ts + brca_metrics$single_mapped.ts
write.csv(brca_metrics, file = "/Users/hchintalapudi/Desktop/work/TCGA analysis/PanCancer/TCGA-BRCA/Telescope_brca_metrics.csv")
brca_metrics<- read.csv("/Users/hchintalapudi/Desktop/work/TCGA analysis/PanCancer/TCGA-BRCA/Telescope_brca_metrics.csv", row.names = 1, stringsAsFactors = F)

# Function for calculating size factors
calculateSizeFactors <- function(mapped_frags) {
    geomean <- expm1(mean(log1p(mapped_frags)))
    mapped_frags / geomean
}
brca_sizefactors<- as.data.frame(calculateSizeFactors(brca_metrics$mapped_frags.bt2))
brca_sizefactors$Sample<- brca_qual_cases$Sample.ID
brca_sizefactors<- brca_sizefactors[,c(2,1)]
colnames(brca_sizefactors)<- c("Sample","sizefactor")
write.csv(brca_sizefactors, file = "/Users/hchintalapudi/Desktop/work/TCGA analysis/PanCancer/TCGA-BRCA/Telescope_brca_sizefactors.csv", row.names = F)
brca_sizefactors<- read.csv("/Users/hchintalapudi/Desktop/work/TCGA analysis/PanCancer/TCGA-BRCA/Telescope_brca_sizefactors.csv")


## Clean up clinical metadata to get survival info:
## found a new metadata file instead of the one ALi gave
brca_clinical_updated<-read.delim("/Users/hchintalapudi/Desktop/work/TCGA analysis/PanCancer/TCGA-BRCA/gdc_download_20220315_152239.658026/8162d394-8b64-4da2-9f5b-d164c54b9608/nationwidechildrens.org_clinical_patient_brca.txt")
brca_clinical_updated<- brca_clinical_updated[-c(1,2),]
brca_clinical_updated<- brca_clinical_updated%>%
  mutate(bcr_patient_uuid = tolower(bcr_patient_uuid))
brca_clinical.2<- brca_clinical_updated[brca_clinical_updated$bcr_patient_uuid %in% brca_qual_cases$id,]
## tumor grade not found!
brca_clinical.2<- brca_clinical.2[,c("bcr_patient_barcode","vital_status","ajcc_pathologic_tumor_stage","age_at_diagnosis",  "last_contact_days_to","death_days_to")]

#brca_clinical.2$age_at_diagnosis[brca_clinical.2$age_at_diagnosis == "[Not Available]"]<- NA
brca_clinical.2<- left_join(brca_clinical.2, brca_qual_cases[,c(2,5)], by = "bcr_patient_barcode")

brca_clinical.2$ajcc_pathologic_tumor_stage[brca_clinical.2$ajcc_pathologic_tumor_stage == "[Discrepancy]" | brca_clinical.2$ajcc_pathologic_tumor_stage == "[Not Available]"]<- NA

brca_clinical.2$ajcc_pathologic_tumor_stage[brca_clinical.2$ajcc_pathologic_tumor_stage == "Stage IA" | brca_clinical.2$ajcc_pathologic_tumor_stage == "Stage IB"]<- "Stage I"

brca_clinical.2$ajcc_pathologic_tumor_stage[brca_clinical.2$ajcc_pathologic_tumor_stage == "Stage IIA" | brca_clinical.2$ajcc_pathologic_tumor_stage == "Stage IIB"]<- "Stage II"

brca_clinical.2$ajcc_pathologic_tumor_stage[brca_clinical.2$ajcc_pathologic_tumor_stage == "Stage IIIA" | brca_clinical.2$ajcc_pathologic_tumor_stage == "Stage IIIB" | brca_clinical.2$ajcc_pathologic_tumor_stage == "Stage IIIC"]<- "Stage III"

brca_clinical.2$ajcc_pathologic_tumor_stage[brca_clinical.2$ajcc_pathologic_tumor_stage == "Stage X"]<- "Stage IV"


brca_clinical.2<- brca_clinical.2%>%
  mutate(censor = ifelse(vital_status == "Alive", 0, ifelse(vital_status == "Dead", 1, NA)))

brca_clinical.2$survival.time <- as.integer(ifelse(brca_clinical.2$death_days_to == "[Not Applicable]",
                                   brca_clinical.2$last_contact_days_to, brca_clinical.2$death_days_to))
brca_clinical.2$survival.time[brca_clinical.2$survival.time<0]<- 0
brca_clinical.3<- brca_clinical.2[,c(7,9,8,4,3)]
## get vst counts:

brca_metadata_deseq<- brca_clinical.3[,c(1,2)]
median(brca_metadata_deseq$survival.time)
rownames(brca_metadata_deseq)<- NULL
brca_metadata_deseq<- brca_metadata_deseq%>%
  mutate(Survival = ifelse(survival.time>=289, "LTS", ifelse(survival.time<=289, "STS",NA)))%>%
  column_to_rownames("Sample.ID")%>%
  dplyr::select(-1)
write.csv(brca_metadata_deseq, file = "/Users/hchintalapudi/Desktop/work/TCGA analysis/PanCancer/TCGA-BRCA/BRCA_metadata_deseq2.csv")
brca_metadata_deseq<- read.csv("/Users/hchintalapudi/Desktop/work/TCGA analysis/PanCancer/TCGA-BRCA/BRCA_metadata_deseq2.csv", row.names = 1)

brca_counts.telescope<- brca_counts.telescope[rownames(brca_metadata_deseq)]
table(colnames(brca_counts.telescope) == rownames(brca_metadata_deseq))

dds_brca<- DESeqDataSetFromMatrix(countData = brca_counts.telescope, colData =brca_metadata_deseq, design = ~Survival)
dds_brca$Survival<- factor(dds_brca$Survival, levels = c("STS", "LTS"))
brca_sizefactors<- brca_sizefactors[match(rownames(brca_metadata_deseq), brca_sizefactors$Sample),]
table(rownames(brca_metadata_deseq) == brca_sizefactors$Sample)

sizeFactors(dds_brca)<- brca_sizefactors$sizefactor
## minimum of 10 counts in 50% of patients:
#dds_lihc.2<- dds_lihc[rowSums(counts(dds_lihc)>=10)>=171,]
dds_brca.2<- dds_brca[rowSums(counts(dds_brca)>=10)>=(ncol(brca_counts.telescope)*0.5),]
table(rowSums(counts(dds_brca)>=10)>= (ncol(brca_counts.telescope)*0.5))

brca_vst<- as.data.frame(assay(vst(dds_brca.2, blind = F)))
write.csv(brca_vst, file = "/Users/hchintalapudi/Desktop/work/TCGA analysis/PanCancer/TCGA-BRCA/brca_vst_0.50_2966.csv")
brca_vst<- read.csv("/Users/hchintalapudi/Desktop/work/TCGA analysis/PanCancer/TCGA-BRCA/brca_vst_0.50_2966.csv", row.names = 1, check.names = F)
```


```{r}
table(colnames(brca_vst) == brca_clinical.3$Sample.ID)

brca_cox_input<- data.frame(brca_clinical.3[,c("Sample.ID","survival.time","censor","age_at_diagnosis", "ajcc_pathologic_tumor_stage")], t(brca_vst))
rownames(brca_cox_input)<- NULL
brca_cox_input<- brca_cox_input%>%
  column_to_rownames("Sample.ID")

brca_cox_input$survival.time<- as.numeric(brca_cox_input$survival.time)
brca_cox_input$censor<- as.numeric(brca_cox_input$censor)
brca_cox_input$age_at_diagnosis<- as.numeric(brca_cox_input$age_at_diagnosis)
brca_cox_input$ajcc_pathologic_tumor_stage<- as.factor(brca_cox_input$ajcc_pathologic_tumor_stage)

write.csv(brca_cox_input, file = "/Users/hchintalapudi/Desktop/work/TCGA analysis/PanCancer/TCGA-BRCA/brca_cox_input_445.csv")

## Cox with RegParalell:
brca_cox_res<-RegParallel(
    data = brca_cox_input,
    formula = 'Surv(survival.time, censor) ~ [*] + age_at_diagnosis + ajcc_pathologic_tumor_stage',
    FUN = function(formula, data)
      coxph(formula = formula,
        data = data,
        ties = 'breslow',
        singular.ok = TRUE),
    FUNtype = 'coxph',
    variables = colnames(brca_cox_input)[5:ncol(brca_cox_input)],
    blocksize = 1000,
    cores = 2,
    nestedParallel = FALSE,
    conflevel = 95,
    excludeTerms = c("age_at_diagnosis", "ajcc_pathologic_tumor_stage"),
    excludeIntercept = TRUE,
    p.adjust = 'fdr')
write.csv(brca_cox_res, file = "/Users/hchintalapudi/Desktop/work/TCGA analysis/PanCancer/TCGA-BRCA/BRCA-vst_cox_results_0.50_2966.csv")
```


```{r}
###--- Gene Counts:
brca_unstr<- read.delim("/Users/hchintalapudi/Desktop/work/TCGA analysis/PanCancer/TCGA-BRCA/TCGA-BRCA_unstr-counts.txt", check.names = F, row.names = 1, stringsAsFactors = F)
brca_unstr<- brca_unstr[samples]
table(samples == colnames(brca_unstr))
table(samples == brca_qual_cases$File.Name)
colnames(brca_unstr)<- brca_qual_cases$Sample.ID
write.csv(brca_unstr, file = "/Users/hchintalapudi/Desktop/work/TCGA analysis/PanCancer/TCGA-BRCA/TCGA-BRCA_unstr-counts_raw_445.txt")



```

